Amazon.IonDotnet.Tests/Internals/TextReaderTest.cs

```diff
@@ -138,16 +138,77 @@ public void FlatIntList()
             ReaderTestCommon.FlatIntList(reader);
         }
 
-        //        [TestMethod]
-        //        public void ReadAnnotations_SingleField()
-        //        {
-        //            // a singlefield structure with annotations
-        //            // {withannot:years::months::days::hours::minutes::seconds::18}
-        //            var annotSingleField = DirStructure.ReadDataFile("text/annot_singlefield.ion");
-        //            var converter = new SaveAnnotationsReaderRoutine();
-        //            var reader = new UserTextReader(new MemoryStream(annotSingleField), converter);
-        //            ReaderTestCommon.ReadAnnotations_SingleField(reader, converter);
-        //        }
+        [TestMethod]
+        public void ReadTypeAnnotations_SingleField()
+        {
+            // a singlefield structure with annotations
+            // {withannot:years::months::days::hours::minutes::seconds::18}
+            byte[] data = DirStructure.OwnTestFileAsBytes("text/annot_singlefield.ion");
+            UserTextReader reader = new UserTextReader(new MemoryStream(data));
+
+            ReaderTestCommon.ReadTypeAnnotations_SingleField(reader);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(UnknownSymbolException))]
+        public void ReadTypeAnnotations_AssertUnknownSymbolException()
+        {
+            byte[] data = DirStructure.OwnTestFileAsBytes("text/unknown_symbols.ion");
+            UserTextReader reader = new UserTextReader(new MemoryStream(data));
+
+            reader.MoveNext();
+            reader.GetTypeAnnotations();
+        }
+
+        [TestMethod]
+        public void ReadTypeAnnotationSymbols_SingleField()
+        {
+            // a singlefield structure with annotations
+            // {withannot:years::months::days::hours::minutes::seconds::18}
+            byte[] data = DirStructure.OwnTestFileAsBytes("text/annot_singlefield.ion");
+            UserTextReader reader = new UserTextReader(new MemoryStream(data));
+
+            ReaderTestCommon.ReadTypeAnnotationSymbols_SingleField(reader);
+        }
+
+        [TestMethod]
+        public void ReadTypeAnnotationSymbols_AssertNoUnknownSymbolException()
```

Comment:
These `UnknownSymbolException` tests are a good start, but in addition we need to test the case when `UnknownSymbolException` is expected to be thrown from `HasAnnotation()`, and we need all three of these test cases for not just the text reader but also the binary and tree reader.

Reply:
Added additional unit tests. One thing to note, I'm having difficulty in writing unknown symbols in binary. I've added a test case that calls `GetTypeAnnotationSymbols` and asserts the exception. The BinaryWriter disallows such unknown symbols.

Reply:
It sounds like you just need to start with the equivalent of
`$ion_symbol_table::{ imports:[{ name: "abc", version: 1, max_id: 1}],symbols: ["foo"]}$10::$11::"value"`
in binary Ion. That's
```
E0 01 00 EA EE 98 81 83 DE 94 86 BC DB 84 83 61 62 63 85 21 01 88 21 01 87 B4 83 66 6F 6F E9 82 8A 8B 85 76 61 6C 75 65 
```
(hexadecimal bytes)

Note: I generated this with the following code that uses ion-java:

```java
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (_Private_IonRawWriter writer = IonBinaryWriterBuilder.standard().build(out).asFacet(_Private_IonManagedWriter.class).getRawWriter()) {
            writer.addTypeAnnotationSymbol(3);
            writer.stepIn(IonType.STRUCT);
            writer.setFieldNameSymbol(6);
            writer.stepIn(IonType.LIST);
            writer.stepIn(IonType.STRUCT);
            writer.setFieldNameSymbol(4);
            writer.writeString("abc");
            writer.setFieldNameSymbol(5);
            writer.writeInt(1);
            writer.setFieldNameSymbol(8);
            writer.writeInt(1);
            writer.stepOut();
            writer.stepOut();
            writer.setFieldNameSymbol(7);
            writer.stepIn(IonType.LIST);
            writer.writeString("foo");
            writer.stepOut();
            writer.stepOut();
            writer.addTypeAnnotationSymbol(10);
            writer.addTypeAnnotationSymbol(11);
            writer.writeString("value");
        }
        for (byte b : out.toByteArray()) {
            System.out.print(String.format("%02X ", b));
        }
```

Create a reader over a stream that contains those bytes, and you should be good to go.

Reply:
Thanks, I've used this snippet to output the contents to a binary data file to follow the same pattern as the other binary tests. I've added the missing test cases.